rbd: do deep copy for dummyVol struct

Madhu-1 opened the pull request:
with a shallow copy of rbdVol to dummyVol, the image name update of the dummyVol is getting reflected on the rbdVol which we don't want.
do a deep copy to avoid this problem.

fixes #https://github.com/ceph/ceph-csi/pull/2656#discussion_r754132353
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: do deep copy for dummyVol struct

with shallow copy of rbdVol to dummyVol
the image name update of the dummyVol is getting
reflected on the rbdVol which we dont want.

do deep copy to avoid this problem.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@BenamarMk PTAL.

Madhu-1 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

Madhu-1 commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.21

`Warning: Permanently added '192.168.39.159' (ECDSA) to the list of known hosts.
client_loop: send disconnect: Broken pipe
script returned exit code 255?
`

BenamarMk reviewed (approved):
Tested this patch and it works as intended.

Madhu-1 commented:
## Before the fix

```log
I1122 09:05:59.458723       1 utils.go:177] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC call: /replication.Controller/EnableVolumeReplication
I1122 09:05:59.459657       1 utils.go:179] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:05:59.502402       1 omap.go:87] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
I1122 09:05:59.581764       1 rbd_util.go:345] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 rbd: create replicapool/csi-vol-dummy-818d066f-39ab-45de-8943-a117f9d82c8e size 1024M (features: [layering]) using mon 192.168.121.235:6789
I1122 09:06:00.662432       1 utils.go:188] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC response: {}
I1122 09:06:00.668961       1 utils.go:177] ID: 20 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC call: /replication.Controller/PromoteVolume
I1122 09:06:00.669184       1 utils.go:179] ID: 20 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:06:00.683420       1 omap.go:87] ID: 20 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
E1122 09:06:00.720896       1 utils.go:186] ID: 20 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC error: rpc error: code = InvalidArgument desc = mirroring is not enabled on 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004, image is in 2 Mode
I1122 09:06:00.751022       1 utils.go:177] ID: 21 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC call: /replication.Controller/EnableVolumeReplication
I1122 09:06:00.756642       1 utils.go:179] ID: 21 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:06:00.763371       1 omap.go:87] ID: 21 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
I1122 09:06:01.672882       1 utils.go:188] ID: 21 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC response: {}
I1122 09:06:01.692753       1 utils.go:177] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC call: /replication.Controller/PromoteVolume
I1122 09:06:01.693003       1 utils.go:179] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:06:01.698206       1 omap.go:87] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
I1122 09:06:01.736907       1 replicationcontrollerserver.go:526] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 Attempting to tickle dummy image for restarting RBD schedules
I1122 09:06:03.784935       1 replicationcontrollerserver.go:538] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 Added scheduling at interval 5m, start time  for volume replicapool/csi-vol-dummy-818d066f-39ab-45de-8943-a117f9d82c8e
I1122 09:06:03.785582       1 utils.go:188] ID: 22 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC response: {}
```

## After the fix
```log
30-0242ac110004 GRPC call: /replication.Controller/EnableVolumeReplication
I1122 09:15:59.930725       1 utils.go:179] ID: 18 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:15:59.960524       1 omap.go:87] ID: 18 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
I1122 09:16:00.033938       1 rbd_util.go:345] ID: 18 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 rbd: create replicapool/csi-vol-dummy-818d066f-39ab-45de-8943-a117f9d82c8e size 1024M (features: [layering]) using mon 192.168.121.235:6789
I1122 09:16:00.036236       1 utils.go:188] ID: 18 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC response: {}
I1122 09:16:00.037846       1 utils.go:177] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC call: /replication.Controller/PromoteVolume
I1122 09:16:00.038253       1 utils.go:179] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC request: {"parameters":{"mirroringMode":"snapshot","schedulingInterval":"5m"},"secrets":"***stripped***","volume_id":"0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004"}
I1122 09:16:00.039916       1 omap.go:87] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 got omap values: (pool="replicapool", namespace="", name="csi.volume.0c25bdd3-485f-11ec-bd30-0242ac110004"): map[csi.imageid:10b183a48a97 csi.imagename:csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004 csi.volname:pvc-26893f08-ff2b-4a0f-a5c3-884b720ffb2c csi.volume.owner:default]
I1122 09:16:00.093362       1 replicationcontrollerserver.go:526] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 Attempting to tickle dummy image for restarting RBD schedules
I1122 09:16:01.622962       1 replicationcontrollerserver.go:538] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 Added scheduling at interval 5m, start time  for volume replicapool/csi-vol-0c25bdd3-485f-11ec-bd30-0242ac110004
I1122 09:16:01.623516       1 utils.go:188] ID: 19 Req-ID: 0001-0009-rook-ceph-0000000000000002-0c25bdd3-485f-11ec-bd30-0242ac110004 GRPC response: {}
```

Madhu-1 commented:
@humblec @nixpanic can I get a review on this one on priority?

Madhu-1 committed:
rbd: operate on dummy image after adding scheduling

currently we are fist operating on the  dummy
image to refresh the pool and then we are adding
the scheduling. we think the scheduling should
be added first and than we should refresh the
pool. If we do this all the existing schedules
will be considered from the scheduler.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
added DNM. QE is still testing this one  looks like we need some more fix

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
>Cannot contact cico-workspace-ptd83: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@6d19570c:JNLP4-connect connection from 10.131.0.145/10.131.0.145:41744": Remote call on JNLP4-connect connection from 10.131.0.145/10.131.0.145:41744 failed. The channel is closing down or has closed down

looks like jenkins issue restarting  the tests

Madhu-1 commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Madhu-1 commented:
/retest ci/centos/mini-e2e/k8s-1.21

humblec commented:
@Madhu-1 if I heard right we are waiting for QE testing, isnt it ?

Madhu-1 commented:
> @Madhu-1 if I heard right we are waiting for QE testing, isnt it ?

yes, added DNM for the same reason.

nixpanic commented:
@agarwal-mudit could you share some of the test results please? Re-adding `DNM` for now.

nixpanic commented on internal/rbd/replicationcontrollerserver.go:
```diff
@@ -290,9 +290,9 @@ func createDummyImage(ctx context.Context, rbdVol *rbdVolume) error {
 			return err
 		}
-		dummyVol := rbdVol
+		dummyVol := *rbdVol
```
It probably helps to have a comment about this, as the implicit deep-copy is something that is really required here. Or create a `DeepCopy()` function to make it very explicit?

Madhu-1 replied:
I can add a comment. if I make any code changes this should go through the testing again to cover all the cases to avoid regression.

nixpanic replied:
Do we have regression tests for this at all? If the e2e does not have a test-case, you can add the `ci/skip/e2e` label?

The current change is very fragile, and breaking seems easy. If a follow-up is planned, can you share a link to the issue?

Madhu-1 replied:
@nixpanic we are running the e2e to make sure nothing breaks at a cephcsi level. Yes, we can add the label here. nothing is planned an as a follow-up as code fix. https://github.com/ceph/ceph-csi/issues/2675  a tracker to remove this workaround

Madhu-1 replied:
regression is a manual testing for now

agarwal-mudit commented:
> @agarwal-mudit could you share some of the test results please? Re-adding `DNM` for now.

It was decided not to wait for the test results. We can merge the fix and if it doesn't help QE can put it back to ON_QA.
Adding you in the discussion. Removing the DNM label

Madhu-1 commented:
@nixpanic @humblec This is tested from @BenamarMk and Annette and its working fine as expected
